[2단계 - 웹 기반 로또 게임] 기린(정유정) 미션 제출합니다.

jeongyou opened the pull request:
동키콩 안녕하세요~! 기린입니다! 2단계 리뷰도 잘 부탁드립니다.

## 학습 목표
이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.
- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트  
- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [ ] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: [https://jeongyou.github.io/javascript-lotto/](url)

## 이번에 작성한 코드의 폴더 구조
```
JAVASCRIPT-LOTTO
│── css                       # 스타일링 관련 폴더
│   │── modal.css             # 모달 스타일링
│   │── style.css             # 페이지 전체 스타일링
│── src                       
│   │── step2-index.js        # 로또 기능 구현
│── index.html                # 로또 웹페이지의 화면 구성
```

## 리뷰 요청 & 논의하고 싶은 내용  
<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.  
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->
### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점
<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? --> 
1. 도메인을 수정하지 않고 웹에서 잘 실행되도록 하려고 했습니다. 기능적인 코드는 문제 없이 붙일 수 있었습니다. 하지만 콘솔과 웹에서 입력받는 방식이 다르다 보니 validate 할 때 오류가 발생하여 콘솔, 그리고 웹에서 모두 사용 가능하도록 코드를 수정하였습니다. 이 과정에서 validate를 구현할 때 입력의 타입과 관계 없이 검증하는 것이 좋겠다고 생각했습니다.  

### 2) 이번 리뷰를 통해 논의하고 싶은 부분  
<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->
1. 코드 작성 순서
    - 도메인과 웹 UI를 연결하는 과정을 처음해보고 잘 몰라서 `UI, 스타일 먼저 전부 구현 -> 도메인과 연결` 이렇게 진행했는데 처음에 UI 요소들을 배치하고 스타일링 하는데 시간이 오래 걸렸지만 도메인 코드가 먼저 완성 되어있었기 때문에 나중에 도메인 코드와 연결하는데는 많은 시간이 들지 않고 수월하게 할 수 있었습니다. 저는 이 방법이 해보니까 괜찮은 것 같은데, UI를 만들고 기능을 바로 붙이는 과정을 반복하는 방식도 사용하는 것 같습니다. 혹시 동키콩은 어떤 방식을 선호하시나요? 저는 제가 했던 방식이 나쁘지는 않았던 것 같은데 만약 도메인 로직이 완성되어있지 않고 새로 만들어야 했다면 지금보다 훨씬 시간이 오래걸렸을 것 같아서 더 선호되는 방법이 있는지 궁금합니다.

2. 파일 분리에 관한 궁금증
    - 위의 폴더 구조를 보면 css 관련 코듣 css 폴더에, js 파일은 src에 index.html은 폴더에 두지 않고 그냥 놔뒀습니다. 이 방법은 괜찮은 건가요?? 뭔가 js 파일은 controller의 역할을 하는 것 같고 index.html은 UI 이렇게 되는 것 같은데 react가 아닌 이런 형식의 프로젝트는 어떻게 폴더 구조를 정해야 하는지 조언을 얻고 싶습니다.

3. let과 관련된 궁금증
    - step2-index.js 파일을 보시면 
    ```
    let price = 0;
    let lottos = [];
    ```
    이 두 값은 여러 함수에서 쓰기 때문에 전역적으로 쓰기 위해 let으로 쓸 수 밖에 없었습니다. 이걸 const로 바꿀 수 있는 좋은 방법이 있는지 궁금합니다. 그리고 사실 요구사항의 '변수 선언시 const 만 사용한다.' 이 말에 담긴 의미를 아직 잘 모르겠고 현재 제가 질문한 상황에 적용이 되는지도 잘 모르겠습니다.

4. 도메인 로직 수정 
    - [refactor: 콘솔과 웹, 둘다 사용가능하도록 재입력 로직 변경](https://github.com/woowacourse/javascript-lotto/commit/1ccac92f01e75619923e8b81b630c7f42e1aebdb)
    - [refactor: 웹과 콘솔 둘 다 사용할 수 있도록 당첨번호, 보너스번호 검증 로직 수정](https://github.com/woowacourse/javascript-lotto/commit/a2312ff6f8ed58454deb827f531f25c07bbfd129)
    - 이렇게 두 로직을 변경하게 되었는데 재입력 로직은 변경하면서 while()을 사용하게 되어 depth 1 조건을 위반하게 되었습니다. depth 1 조건을 지키면서도 웹과 콘솔 둘 다 사용할 수 있게 코드를 구성할 수 있는 방법이 있는지 궁금합니다. 
    - 그리고 저는 원래 콘솔에서 검증 할 때 공백을 검증하기 위해서 문자열 그대로 검증하고 검증을 통과하면 number나 배열로 바꿔주었습니다. 그런데 이번 미션을 하면서 숫자로 바로 입력이 들어와 버려서 검증 로직을 고칠 수 밖에 없었습니다. 보통 콘솔에서도 입력을 받은 후에 바로 number나 배열로 변환을 한 후에 검증을 하는지 아니면 검증이 끝나면 그 때 데이터를 변환하는지 궁금합니다.
     ex)
        - 저는 평소에 '6' 입력 -> 공백입력 검증, 숫자로 변환해서 여러가지 검증 -> 검증을 통과하면 `const number = Number('6')`으로 할당  이 방식으로 구현했었는데 
        - '6' 입력, 공백입력 검증, 숫자로 변환 -> 여러가지 검증 -> 검증을 통과하면 `const number = 6` 으로 할당 이런 방식이 맞는건가요...?

5. 저는 계속해서 변경이 될 수 밖에 없는 로또 티켓 생성 부분, 그리고 당첨 통계 모달의 숫자 부분만 동적으로 변경하고 나머지는 정적으로 구현해 화면에 보이지 않아야 할 때는 css에서 숨김처리 해줬습니다. 이 방법이 적절한지 아니면 css에서 숨김처리하지 말고 동적으로 요소를 삭제했다 새로 만드는게 더 적절한지 궁금합니다.

6. HTML에서 class이름을 지을 때 어떻게 해야하는지 잘 모르겠습니다. 규칙 같은 것이 있는지 궁금합니다. 그리고 id를 언제 사용해야하는지도 궁금합니다. 만약 유일하게 존재해야하는 요소라면 무조건 id로 지정해주는 것이 좋나요? class 이름을 다른 것과 겹치게 안하면 id와 똑같은 기능을 하는 것 같아서 id의 역할이 뭔지 잘 모르겠습니다. 

끝입니다 감사합니다.

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계
- [ ] TDD를 활용해 기능을 구현하는 과정에서 적절한 테스트 우선 접근 방식을 적용했는가? 단위 테스트의 커버리지는 충분한가?
- [ ] 도메인과 UI의 관심사를 분리하여 적절한 모듈화가 이루어졌는가? 하나의 객체나 모듈이 너무 많은 책임을 가지고 있지는 않은가?
- [ ] 객체의 프로퍼티를 직접 조작하기보다 메시지를 던지고 있는가?
- [ ] 불필요한 클래스를 사용하지 않고, 함수를 적극적으로 활용하여 JavaScript다운 방식으로 로직을 구현했는가?

### 2단계 
- [ ] 도메인 로직에 불필요한 영향을 주지 않고 UI 변경에 대응했는가?
- [ ] DOM 조작과 이벤트 활용을 JavaScript의 개념에 맞게 이해하고 적절하게 적용했는가? 
- [ ] 웹 표준을 준수하는 마크업을 활용하며, 스타일 작성에 일관성이 있는가?
</details>

jeongyou committed:
docs: 웹 기반 로또 게임 구현 기능 목록 작성

jeongyou committed:
style: 공통 css 지정

jeongyou committed:
feat: 헤더에 스타일 적용

jeongyou committed:
feat: 푸터 적용, 게임이 진행 될 div 생성

jeongyou committed:
feat: 구입 전  UI 마크업 및 스타일 추가

- 타이틀, 구입할 금액을 입력해주라는 텍스트, 구입 금액 입력 칸, 구입 버튼을 추가하고 전체적인 스타일을 적용함.
- README 기능 목록도 업데이트

jeongyou committed:
feat: 로또 구입 후 UI 마크업 및 스타일 추가

- 구입한 로또 개수 및 번호 출력, 당첨 번호 입력 UI를 추가함.
- 당첨 번호 및 보너스 번호 입력 칸을 구성하고, 결과 확인 버튼을 추가함.

jeongyou committed:
style: header 위치 고정하고 container 최대한 안 잘리게 수정

jeongyou committed:
feat: 당첨 통계 모달 UI 마크업 및 스타일 추가

- 당첨 모달 레이아웃 마크업 작성
- `X` 버튼 추가
- 당첨 통계 타이틀 및 표 스타일링
- 수익률 텍스트 및 다시 시작 버튼 추가

jeongyou committed:
docs: 기능 구현 목록 수정

jeongyou committed:
style: 금액, 당첨 번호, 보너스 번호 input에 숫자만 받을 수 있게 함

- style.css 파일에서 input type number 우측 화살표도 제거

jeongyou committed:
refactor: 콘솔과 웹, 둘다 사용가능하도록 재입력 로직 변경

jeongyou committed:
feat: 로또 구입 후 UI 구조 변경 및 초기 숨김 처리

- 구입 금액 입력 폼을 `<form>` 태그로 감싸서 구조 변경
- 구입 후 표시될 UI를 `.lotto-display-container`로 묶음
- `.lotto-display-container`를 추가하고, 기본적으로 `display: none;`으로 설정

jeongyou committed:
feat: 로또 구입 기능 및 UI 업데이트 추가

- 구입 금액 입력 후 로또 번호를 생성하는 기능 추가
- 입력값 검증 후 유효하지 않은 경우 alert로 에러 메시지 출력
- 로또 개수 및 번호 리스트 UI 동적 업데이트 (`updateLottoUI()` 추가)
- 구입 후 UI (`.lotto-display-container`)가 보이도록 처리
- 입력 후 입력 칸 초기화 (`#purchase-input.value = '';`)
- 에러 발생 시 `.lotto-display-container`를 숨기도록 처리

jeongyou committed:
refactor: 웹과 콘솔 둘 다 사용할 수 있도록 당첨번호, 보너스번호 검증 로직 수정

jeongyou committed:
style: 당첨 번호 보너스번호를 form 태그로 감싸도록 UI 마크업 수정

jeongyou committed:
feat: 당첨 번호 입력 및 당첨 통계 기능 추가

- 당첨 번호 6개 및 보너스 번호 1개 입력 기능 구현
- 입력된 당첨 번호 및 보너스 번호 검증 (alert 메시지 출력)
- 결과 확인 버튼 클릭 시 당첨 통계 계산 및 수익률 계산 기능 추가
- 당첨 통계 결과를 모달 UI로 표시하도록 구현

jeongyou committed:
style: 모달의 닫기 버튼에서 x 부분만 눌렀을 때 닫도록 수정

- 버튼 자체의 width가 100%로 설정 되어있어서 x가 아닌 옆 부분을 눌렀을 때도 모달이 닫하는 문제 개선

jeongyou committed:
style: 모달 요소간 margin 조정

jeongyou committed:
style: 당첨 통계 테이블에 3개 당첨 행 2개 중복 중 1개 삭제

jeongyou committed:
feat: 당첨 통계 모달 및 수익률 표시 기능 추가

- `X` 버튼 및 ESC 키 입력 시 모달 닫기 기능 구현
- 당첨 번호 및 보너스 번호 입력값 초기화 기능 추가
- 당첨 통계(일치 개수, 당첨금, 당첨 갯수) 동적으로 업데이트
- 수익률을 소수점 1자리까지 반올림하여 표시 및 1000단위 콤마 추가

jeongyou committed:
feat: 다시 시작하기 버튼 클릭 시 초기화 기능 추가

- "다시 시작하기" 버튼 클릭 시 구입 전 화면으로 초기화
- 게임 데이터(price, lottos) 초기화
- 로또 개수 및 당첨 통계 UI 초기화
- 입력된 구입 금액, 당첨 번호, 보너스 번호 초기화
- 모달 닫기 기능 추가

jeongyou committed:
docs: 목록에 구현 기능 추가

- 구입 후 구입 버튼 비활성화 기능 구현 목록 추가

jeongyou committed:
feat: 구입 버튼 비활성화 및 스타일 변경

- 로또 구입 후 구입 버튼을 비활성화하도록 변경
- 비활성화된 버튼 스타일 추가 (배경 회색, 커서 변경)
- "다시 시작하기" 클릭 시 버튼이 다시 활성화되도록 구현

jeongyou committed:
style: 모든 버튼에 커서 스타일 변경

- 모든 버튼 위에 마우스를 올리면 커서를 `pointer`로 변경

jeongyou committed:
chore: 불필요한 console.log 삭제

jeongyou committed:
refactor: 로또 UI 업데이트 함수 분리

- updateLottoUI 함수의 역할을 세분화하여 함수 분리
  - updateLottoCount: 로또 개수 업데이트
  - renderLottoTickets: 로또 번호 리스트 렌더링
  - createLottoItem: 개별 로또 아이템 생성

jeongyou committed:
refactor: UI 업데이트 로직을 updatePurchaseUI 함수로 분리

- UI 업데이트 로직을 updatePurchaseUI 함수로 분리하여 재사용
- 구매 완료 여부를 isPurchased 인자로 전달
- 불필요한 중복 코드를 제거

jeongyou committed:
refactor: 당첨 번호 검증 및 결과 처리 함수 분리

jeongyou committed:
docs: readme 파일 이름 변경

jeongyou committed:
chore: css 폴더에 css파일 이동

JUDONGHYEOK commented on css/style.css:
```diff
@@ -0,0 +1,443 @@
+  border-spacing: 0;
+}
+
+:root {
```
현재 reset 코드와 여러 스타일 코드가 한 파일에 있어서 스타일코드를 파악하기가 쉽지 않네요. 적절한 관심사에 따라 스타일 코드를 분리해주면 어떨까요?

jeongyou replied:
관심사에 따라서 css를 분리 해 보았습니다.
[refactor: css 파일을 역할별로 분리하고 index.css에서 통합 관리](https://github.com/woowacourse/javascript-lotto/pull/395/commits/d6e335cc8db2283bbdde43159c44d88332d529cf)

- reset.css: 브라우저 기본 스타일 초기화
- variables.css: 전역 변수 (색상, 폰트 크기 등) 관리
- typography.css: 텍스트 스타일 (제목, 본문, 캡션 등) 관리
- layout.css: 페이지 전체 레이아웃 스타일 관리
- components.css: 개별 UI 요소 (버튼, 입력 필드 등) 스타일 추가
- modal.css: 모달 창 스타일 별도 관리
- index.css: 개별 CSS 파일을 불러와 한 번에 적용하도록 수정

❓한 가지 궁금한 점이 있습니다. 현재 modal.css를 개별 파일로 분리했는데, 이렇게 따로 두는 것이 적절할까요? 아니면 layout.css 또는 components.css에 포함하는 것이 더 나을까요?
모달을 독립적으로 관리하는 것이 유지보수에 더 효과적일 것 같긴 하지만, 관심사에 따라 분리된 것이 맞는지 고민이 됩니다.
또 일반적으로 CSS 파일을 역할별로 분리하는 방식이 많이 사용되는 것 같은데,
이런 파일 구분에 대한 컨벤션이나 명확한 기준이 있는지 궁금합니다.

JUDONGHYEOK replied:
훨씬 css구조가 깔끔해져 파악하기 쉬워진 것 같아요! 감사합니다 기린!

modal.css같은 경우는 정답이 없는 것 같아요. 모달의 스타일이 많고 애플리케이션에서 핵심 역할을 한다고 여기신다면 분리하셔도 좋고 아니라면 컴포넌트에 위치시켜도 좋을 것 같아요. 개인적인 생각으로는 component에 위치시켜도 무방하다고는 생각하지만 개인적인 취향차이정도라고 생각합니다!

일반적으로 CSS파일 같은 경우는 공통적인 CSS설정의 경우에는 한곳에 응집시켜두고 보통 컴포넌트와 같은 폴더에 위치시키는 경우가 많아요. css도 결국엔 하나의 모듈이기 때문에 사용되는 파일 근처에 위치시키는 것이 관리하기가 용이하기 때문입니다.

파일 구분에 대한 기준은 팀 컨벤션이나 프로젝트에 따라 달라지기 때문에 명확한 기준이라고 말할만한 것은 없는 것 같습니다. 상황에 따라 발전되가는 것들도 필요하구요. 저도 사용해본적 없긴 하지만 SMACSS나 ITCSS같은 방법론들도 존재하니 필요하시면 참고하셔도 좋을 것 같아요

JUDONGHYEOK commented on src/step2-index.js:
```diff
@@ -1,4 +1,170 @@
+const closeBtn = document.querySelector('.close-btn');
+const purchaseBtn = document.querySelector(".purchase-btn");
+let price = 0;
+let lottos = [];
```
해당 내용들을 전역상태로 관리하는 것이 코드의 흐름을 방해할 수 있을 것 같아요. 또 상태변경도 추적하기가 힘들구요. 해당 상태가 step2-index.js에서 관리되어야 하는 것인지 고민할 필요가 있어보여요. 해당 상태의 위치나 캡슐화에 대해서 개선할 수 있을 것 같습니다!

jeongyou replied:
도메인과 UI 객체를 분리하는 과정에서, LottoGame 클래스가 컨트롤러 역할을 하도록 변경하고 해당 클래스에서 상태를 관리하도록 개선했습니다.
상태 변경도 LottoGame 내부에서 추적할 수 있도록 리팩토링했습니다.
[도메인 로직과 UI 로직 분리하여 코드 구조 개선](https://github.com/woowacourse/javascript-lotto/pull/395/commits/d1dde28feef8452f440006c42da5983bcf1e6305)

❓제가 리뷰의 의도를 제대로 이해하고 리팩터링 한 것이 맞나요...?

JUDONGHYEOK replied:
오 컨트롤러를 추가적으로 도입하셨군요! 캡슐화 자체는 이뤄진 것 같지만 조금 더 발전할 수 있을 것 같아요!

> 기린은 Controller가 어떤 역할을 한다고 생각하시나요❓🧐

컨트롤러라고 표현한  LottoGame 의 생성자에서 기본값을 세팅해주고, 검증로직이 들어가 있는 점에서 도메인의 로직과 상태에 직접적인 의존성이 생기고 있어요 😲

 직접 관리의 영역까지 이뤄지게 된다면 컨트롤러는 더욱더 비대해질것이고 도메인과 컨트롤러의 경계는 희미해져 유지보수가 어려워질 것이라고 생각합니다. 
 
 **따라서 지금 상황에서 개선을 한다면 직접 상태와 로직을 갖고 있는 중간 도메인 객체를 생성하시는 것은 어떨까요?**
 컨트롤러는 검증된 입력을 도메인 모델에 전달하고 결과를 UI에 반영하는 역할만 수행하구요. 그렇게 된다면 조금더 계층의 경계가 명확해지지 않을까 생각이 됩니다 👍

jeongyou replied:
저는 처음에는 컨트롤러가 도메인 로직도 직접 수행하는 거라고 생각했습니다. 하지만 코드를 리팩토링하면서, 컨트롤러의 역할은 도메인 로직을 처리하는 게 아니라 검증된 입력을 도메인 모델에 전달하고, 그 결과를 UI에 반영하는 것이라는 걸 깨달았습니다. 기존의 LottoGame에서는 컨트롤러가 Winning을 직접 다루면서 도메인의 상태까지 관리하고 있었는데, 이 부분이 컨트롤러의 역할을 벗어난다고 판단했어요. 그래서 중간 도메인 객체로 LottoGameManager를 추가해서 컨트롤러는 UI와 데이터 흐름을 조율하는 역할만 하고, 핵심 로직은 LottoGameManager에서 처리하도록 개선해보았습니다!

- [refactor(LottoGame): LottoGameManager를 추가하여 상태 및 로직 분리](https://github.com/woowacourse/javascript-lotto/pull/395/commits/621378aa0135998d8c151e9123214af201e216e4)
- [refactor: LottoGame이 UI와 입력값 검증만 담당하도록 리팩토링](https://github.com/woowacourse/javascript-lotto/pull/395/commits/c65546e2ef939218db97358735465f876ae22787)

JUDONGHYEOK replied:
오 좋습니다 저도 기린의 의견에 동의해요. 컨트롤러는 도메인과 ui의 세부 로직에 대해서 모르고 있어야한다고 생각해요. 확실히 중간 도메인 객체를 추가하니 Controller내부의 로직들이 깔끔해진 느낌입니다.

조금 더 개선할 수 있는 부분은 아직 절차지향적인 로직들이 남아있다는 점인데요. 간단한 resetGame 로직만 살펴봐도 UI 처리 과정에서 문제가 드러납니다. 컨트롤러가 UI에게 resetGameUI → enablePurchaseButton → closeResultModal 순으로 3개의 메시지를 보내고 있는데, 이는 이미 UI의 세부 동작에 의존하고 있다는 신호라고 생각해요.
이런 의존성은 결국 UI의 세부 로직에 컨트롤러가 강하게 결합되어 있음을 의미하며, UI 객체 내부에서 더 많은 캡슐화가 필요함을 나타낸다고 생각해요. 물론 어느 정도의 트레이드오프는 불가피하겠지만, "컨트롤러가 정말 이 모든 세부 정보를 알아야 하는가?"라는 질문에 대해 한번 더 생각해볼 필요가 있을 것 같습니다.
 제의견은 의견일뿐이고 다양하게 적용해가며 코드에서 기린만의 생각을 만들어나가면 좋을 것 같아 참고차 남겨봅니다!

JUDONGHYEOK commented on src/step2-index.js:
```diff
@@ -1,4 +1,170 @@
+    updatePurchaseUI(false);
+    alert(error.message);
+  }
+});
```
이런 로직들에서 Dom조작 로직과 비즈니스 로직이 혼재되어 있는 것 같아요. 현재 step2-index.js파일에서 너무 많은 정보를 알고 있어야하고 이는 곧 변경에 취약한 설계라는 것을 의미합니다. 적절하게 도메인과 UI 객체를 분리해서 객체에게 메세지를 보내는 형식으로 변경해 보는 것은 어떨까요?

jeongyou replied:
도메인과 UI 객체를 분리하고, 객체 간 메시지를 주고받는 방식으로 리팩터링해 보았습니다.
[refactor: 도메인 로직과 UI 로직 분리하여 코드 구조 개선](https://github.com/woowacourse/javascript-lotto/pull/395/commits/d1dde28feef8452f440006c42da5983bcf1e6305)

- LottoGame을 컨트롤러 역할로 변경하여 UI와 도메인 로직을 연결했습니다.
- LottoUI는 DOM 조작 및 UI 업데이트를 담당하도록 분리하였습니다.
- step2-index.js에서는 LottoGame을 호출하는 역할만 수행하도록 변경하였습니다.
- step2-index.js에서는 이벤트 리스너만 관리하도록 역할을 최소화하였습니다.
- 전역 상태를 제거하고, 도메인과 UI 객체를 분리하여 코드의 응집도를 높였습니다.

❓각 객체가 자신의 역할만 수행하도록 개선해 보았는데, 한 가지 궁금한 점이 있습니다.
현재 LottoGame에서 UI의 함수뿐만 아니라 Validate, LottoMachine, Winning까지 직접 사용하고 있는데, 이렇게 많은 의존성을 가지는 것이 과도한 역할을 하고 있는 것은 아닌지 고민이 됩니다. 이 부분을 더 개선할 수 있는 방법이 있을까요?

JUDONGHYEOK replied:
> ❓각 객체가 자신의 역할만 수행하도록 개선해 보았는데, 한 가지 궁금한 점이 있습니다.
> 현재 LottoGame에서 UI의 함수뿐만 아니라 Validate, LottoMachine, Winning까지 직접 사용하고 있는데, 이렇게 많은 의존성을 가지는 것이 과도한 역할을 하고 있는 것은 아닌지 고민이 됩니다. 이 부분을 더 개선할 수 있는 방법이 있을까요?

앞선 리뷰와 동일하게 이어지는 내용인 것 같아요! 기린이 느낀바와 같이 LottoGame에 많은 의존성이 있는 것 같습니다. 컨트롤러는 단순히 중간에서 메세지를 통해 모델과 view사이에서 중재자역할만 하는 것이 좋을 것 같아요!

 의존성이라면 따라오는 결합도라는 것도 생각해볼 필요가 있을 것 같습니다!!

JUDONGHYEOK commented on src/step2-index.js:
```diff
@@ -1,4 +1,170 @@
+  lottos.forEach((lotto) => {
+    lottoList.appendChild(createLottoItem(lotto));
+  });
+}
```
DOM조작은 매우 비용이 비싼 행위인데요. 비싼행위가 불필요하게 반복되고 있는 것 같아요! 개선이 필요해보입니다!

jeongyou replied:
DOM 조작을 최소화 하도록 개선해보았습니다.
[refactor: Lotto 목록 렌더링 최적화하여 DOM 조작 최소화](https://github.com/woowacourse/javascript-lotto/pull/395/commits/5ee1a4f14abd2b3cd3104c7d27eb2f0061176f60)

`innerHTML = ""` 후 `appendChild()` 반복해서 호출하는 대신 `map().join("")` 방식으로 변경하여 전체 리스트를 한번에 업데이트 하도록했습니다.

❓innerHTML을 사용하면 XSS(크로스 사이트 스크립팅) 공격에 취약할 수 있다는 내용을 봤습니다. 하지만 로또 번호를 생성하는 경우에는 사용자 입력과 관련이 없기 때문에, 저는 innerHTML을 사용해도 안전하다고 판단하여 적용했습니다.

다만, 혹시 모를 보안 이슈를 대비해 innerHTML 대신 textContent를 사용하는 것이 더 안전한 방법인지 궁금합니다. 이 경우에도 innerHTML을 지양해야 할까요?

또한, 처음 코드처럼 요소들을 textContent를 사용해 생성하면 innerHTML을 사용할 수 없기 때문에, 대신 createDocumentFragment()를 활용하여 한 번만 DOM을 업데이트하는 방식도 고려해볼 수 있을 것 같습니다. 지금 리펙터링 한 코드는 XSS 공격과는 직접적인 관련이 없어 보이는데, 현재 방식 그대로 사용해도 괜찮을까요?

JUDONGHYEOK replied:
오 보안까지 신경써주시다니 아주 나이스입니다~😁

현재 케이스에서는 innerHTML을 사용해도 안전합니다. XSS 취약점이 발생하는 경우는 주로 사용자 입력이나 외부 소스의 데이터를 검증 없이 DOM에 삽입할 때인데, 로또 번호는 애플리케이션 내부에서 생성되는 신뢰할 수 있는 데이터이기 때문이에요. 일반적으로는 말씀해주신대로 textContent나 Fragment, DOMParser등을 활용할 수 있을 것 같은데요.

 현재는 앞서 말씀드린바와 같이 애플리케이션 내부에서 생성된 신뢰할 수 있는 데이터만 사용하기 떄문에 코드 가독성과 유지보수성 측면에서 템플릿 리터럴과 innerHTML을 사용하는 것도 좋아 보여요!!😃

jeongyou replied:
textContent, Fragment, DOMParser과 같은 것들도 있군요! 현재는  innerHTML을 사용해도 될 것 같아 수정하지는 않겠습니다!

JUDONGHYEOK replied:
넵 좋습니다! 현재는 innerHTML로도 충분한 것 같아요!

JUDONGHYEOK commented on src/Controller/LottoController.js:
```diff
@@ -48,35 +48,41 @@ class LottoController {
-      return this.#readWinningNumbers();
+    while (true) {
+      try {
+        const winningNumbers = await InputView.readWinningNumbers();
+        Validate.checkIsEmpty(winningNumbers);
+        const winningNumbersArray = winningNumbers.split(',');
+        Validate.validateWinningNumbers(winningNumbersArray);
+        return winningNumbersArray.map(Number);
+      } catch (error) {
+        OutputView.printErrorMessage(error.message);
+      }
```
이전 리뷰에서도 남긴 것 같지만 validate가 controller에서 모두 이뤄지는 것이 Controller가 비대해지는데 기여를 하고 있지 않나 생각이 돼요. 도메인 객체가 자신의 책임 범위 내에서 검증을 수행하도록 하는 것이 좋을 것 같아요.

jeongyou replied:
❓리뷰해주신 내용을 고민해봤는데, 적용 방법이 잘 감이 오지 않아 질문드립니다!

말씀해주신 대로 Controller에서 검증을 수행하는 것이 아니라, 도메인 객체(Winner, Lotto 등)에서 자체적으로 검증을 수행하는 것이 맞다는 점은 이해했습니다. 하지만 현재 코드에서는 Validate를 Controller에서 호출하고 있고, 이거를 도메인 내부로 이동하려면 코드 구조를 많이 변경해야 할 것 같습니다.

도메인 객체가 스스로 검증하도록 만들려면,
Controller에서 검증 후 객체를 생성하는 방식 → 유효한 값만 생성할 수 있도록 도메인 내부에서 검증하는 방식으로 변경해야 할 것 같은데,
이렇게 변경할 경우 Controller의 역할을 어디까지 줄여야 하는지, 그리고 어떤 방식으로 적용하는 것이 적절할지 고민이 됩니다.
혹시 현재 코드 구조를 크게 변경하지 않으면서 적용할 수 있는 방법이 있을까요?
조언해주시면 감사하겠습니다!

JUDONGHYEOK replied:
컨트롤러 관련 내용이 많네요. 동일한 리뷰를 계속 남기고 있는 것 같아 괜히 죄송해지네요😭

앞선 말과 더불어 드리고 싶은 말씀은 객체의 역할과 결합도에 대해서 다시한번 생각해보면 좋을 것 같아요. 객체지향프로그래밍에서 기본이 되는 원칙 중 하나라고 생각되는 점이 캡슐화인데요. 캡슐화를 이루기 위한 조건중 하나가 객체는 연관된 작업만을 수행하고 연관성이 없는 작업은 다른 객체에게 위임하는 것이 필요하다고 생각해요. 현재 코드에서 많은 의존성들이 객체 간에 존재하고 또 존재하는 지점이 컨트롤러에 위치하고 있다고 생각해요. 따라서 이러한 의존성들을 정리하고 적절하게 객체간의 역할을 위임하는 것이 필요해보여요!

추가로 제가 설명드리고 싶은내용이 오브젝트 라는 책의 1장에 있는 것 같아요! 해당 내용을 ([정리한 글](https://incheol-jung.gitbook.io/docs/study/object/2020-03-10-object-chap1))을 공유드려요!!(직접 책을 읽어보시는 것을 추천드립니다!)

jeongyou replied:
좋은 글 공유해 주셔서 감사합니다! 😊 덕분에 객체가 연관된 작업만 수행해야 한다는 점과, 컨트롤러의 역할을 더 명확하게 이해할 수 있었습니다. 말씀해 주신 것처럼 의존성을 정리하고, 객체 간의 역할을 적절히 위임하는 것이 중요하다는 걸 다시 한번 깨달았습니다. 리뷰를 계속 남겨주시는 게 오히려 저한테는 많은 도움이 되고 있어요! 감사합니다! 🙌

JUDONGHYEOK reviewed (changes requested):
안녕하세요 기린! 2단계도 너무 수고많으셨습니다. 구조적인 개선이 우선적으로 이뤄지면 좋을 것 같아 많은 리뷰를 남기기 보다는 빠른 Reqeust Chages를 요청드려요! 특히 step2-index.js의 에서 도메인 로직과 UI로직이 강하게 결합되어 있는 것 같아요. step2에서 너무 많은 책임을 갖고 있는 것 같아요. Domain은 세부구현사항을 너무나 많이 알고 있어야하고 view에서는 view 로직에 대해서  모듈단위로 분리가 잘 되어 있는지 확인해보시면 좋을 것 같아요!

간단하게 질문에 답변을 드려보겠습니다!
## 1. 코드작성순서
 기린님이 말씀하신 것처럼 도메인 코드가 이미 완성되어 있어서 연결 과정이 수월했다고 하셨는데, 맞아요. 만약 도메인 로직이 아직 완성되지 않았다면 UI를 먼저 만들었을 때 나중에 로직 변경에 따라 UI도 계속 수정해야 하는 상황이 생길 수 있었다고 생각해요.

 도메인 로직을 먼저 작성하고 UI를 나중에 붙이는 방식은 핵심 비즈니스 로직을 안정적으로 다듬을 수 있고, UI가 도메인에 종속되지 않아 더 유연한 구조를 만들기 좋다고 생각해요. 때문에 저는 개인적으로 도메인 로직을 먼저 다듬는 방식을 선호합니다!

 UI를 먼저 구현하는 것도 물론 직관적이고 사용자 경험을 빠르게 확인할 수 있다는 장점이 있지만 대부분의 상황에서는 도메인을 먼저 구현하는 것에 장점이 크다고 생각합니다.
## 2. 파일분리
현재 CSS는 css 폴더에, JS는 src에, index.html은 루트에 두셨다고 하셨는데, 이 구조 자체는 깔끔하고 괜찮아요. 소규모 프로젝트에서는 충분히 적절할 수 있습니다. 다만, step2-index.js가 너무 많은 역할을 하고 있는 점이 걱정돼요. 말씀하신 것처럼 JS가 컨트롤러 역할을 하고 index.html이 UI를 담당하는 느낌이라면, 현재 구조에서는 앞서말씀드린대로 JS 파일을 더 세분화해서 모듈화하는 게 좋을 것 같아요.

## 3. let사용
let사용은 필요에 따라 이뤄질 수도 있겠지만 현재 구조에서는 위험하다고 생각해요. 요구사항에서 "변수 선언 시 const만 사용한다"고 한 이유는 코드의 안정성과 예측 가능성을 높이기 위해서라고 생각해서 이부분은 개선이 필요할 것 같습니다. 현재는 언제 어디에서 데이터가 변경되는지 파악하기 쉽지 않고, 유지보수의 측면에서도 실수를 일으킬 확률이 매우 높아보입니다. 

개선을 하기 위해서는 객체를 역할에 따라 분리해서 적절한 캡슐화가 필요해보여요! 객체를 적절히 잘 활용한다면 변경시점도 명확해지고 가독성도 좋아져 유지보수성을 높일 수 있지 않을까 생각합니다.

## 4. 도메인 로직 수정
 우선, 검증과 형변환 순서에 대해서는 검증의 목적에 따라 다르게 접근할 수 있을 것 같아요. 입력값의 형식(공백, 문자열 패턴 등)을 검증하는 것은 UI 계층의 책임이므로 형변환 전에 수행하고, 비즈니스 규칙(숫자 범위, 중복 여부 등)을 검증하는 것은 도메인 계층의 책임이므로 형변환 후에 수행하는 것이 자연스러울 것 같아요.

 while문으로 인한 depth 1 위반 문제는 다양하게 해결할 수 있을 것 같아요. 기존에 수행하셨던 재귀로도 가능하고 또 과하긴 하지만 이벤트 기반 구조나 플래그변수를 활용할 수 있을 것 같구요. 적절한 트레이드 오프에 따라 구성하면 좋을 것 같지만 제 개인적인 의견으로는 현재도 충분히 좋은 구조라 생각이 됩니다.


## 5. css 숨김처리
 로또 티켓 생성이나 당첨 통계 모달처럼 동적으로 변하는 부분만 JS로 처리하고, 나머지는 CSS로 숨김 처리한 방식은 비용 측면에서 효율적이고 괜찮아요. 다만, CSS로 숨긴 요소가 DOM에 계속 남아 있으면 보안상 노출될 가능성이 있거나 불필요한 메모리를 차지할 수 있다는 단점도 있어요. 반대로 동적으로 요소를 삭제하고 생성하면 코드가 복잡해질 수 있지만, DOM이 깔끔해지고 노출 위험이 줄어듭니다. 현재 규모라면 CSS 숨김 처리로 충분해 보이니, 프로젝트가 커질 때 동적 생성을 고려해보시면 좋을 것 같습니다!

## 6. class, id 
 클래스의 이름을 짓는 것은 은근히 까다로운 일이라고 생각해요. BEM과 같이 여러 클래스이름을 짓는 규칙들이 있으니 참고해보면 좋을 것 같습니다!

 클래스와 id를 사용해야하는 때는 잘 알고 계실거라고 생각해요. class는 재사용 가능한 스타일이나 행동을 그룹화할 때 사용하고 id는 페이지에서 유일한 요소를 식별할 때 사용하죠. 모든 유일한 요소에 id를 넣기보다는 유일한 요소여야하고 이를 식별해야할 때 id를 사용해야하는 것 같아요.

 class가 아닌 id를 넣어준 것 자체가 해당 요소는 고유성을 보장해야한다는 의미를 담고 있기 때문에 해당 요소가 어떠한 의미를 가지느냐를 생각하고 적절한 사용을 하면 좋지 않을까 생각합니다.


---
핵심은 step2-index.js의 책임 분리와 전역 상태 관리 개선일것 같습니다! 위 피드백을 반영해서 코드를 조금 손보면 훨씬 깔끔해질 것 같습니다. 추가 질문 있으면 언제든 말씀해주세요! 다시 한 번 수고 많으셨어요 기린!!!🦒

jeongyou committed:
refactor: css 파일을 역할별로 분리하고 index.css에서 통합 관리

- reset.css: 브라우저 기본 스타일 초기화
- variables.css: 전역 변수 (색상, 폰트 크기 등) 관리
- typography.css: 텍스트 스타일 (제목, 본문, 캡션 등) 관리
- layout.css: 페이지 전체 레이아웃 스타일 관리
- components.css: 개별 UI 요소 (버튼, 입력 필드 등) 스타일 추가
- modal.css: 모달 창 스타일 별도 관리
- index.css: 개별 CSS 파일을 불러와 한 번에 적용하도록 수정

jeongyou committed:
refactor: 도메인 로직과 UI 로직 분리하여 코드 구조 개선

- LottoGame을 컨트롤러 역할로 변경하여 UI와 도메인 로직 연결
- LottoUI에서 DOM 조작 및 UI 업데이트 담당
- step2-index.js에서 LottoGame을 호출하는 역할만 수행하도록 변경
- step2-index.js에서 이벤트 리스너만 관리하도록 역할 분리
- 전역 상태 제거 및 도메인과 UI 객체 분리

jeongyou committed:
refactor: Lotto 목록 렌더링 최적화하여 DOM 조작 최소화

- `innerHTML = ""` 후 `appendChild()` 반복 호출 대신 `map().join("")` 방식으로 변경
- 전체 리스트를 한 번에 업데이트해서 DOM 조작 최소화

jeongyou commented:
안녕하세요 동키콩🦍! 기린입니다~🦒. 꼼꼼한 리뷰 정말 감사합니다!
각 리뷰 코멘트에 대한 수정 사항을 반영하고, 해당 댓글에 코멘트를 남겨놓았습니다!

혹시 추가로 개선할 부분이 있다면 피드백 부탁드립니다. 🙌

JUDONGHYEOK reviewed (changes requested):
안녕하세요 기린!! 빠르게 리뷰를 반영해주셨네요 감사합니다!!☺️
기본적인 요구사항들은 만족하신 것 같은데요. 아직 기린과 객체의 역할에 대해 나눌 이야기가 많은 것 같아 RequestChanges하도록 하겠습니다!

모든 리뷰를 반영해주시지 않으셔도 되고 기린만의 생각을 코멘트로 남겨주셔도 되니 편하게 말씀해주시면 감사하겠습니다!!🙇🏻‍♂️

jeongyou committed:
refactor(Winning, Validate, LottoController): 객체 간 역할 정리 및 검증 책임 위임

- Winning이 당첨 번호 및 보너스 번호 검증을 직접 수행하도록 리팩터링
- Validate를 유틸성 함수 모음으로 단순화하여 검증을 Winning 내부에서 호출
- LottoController에서 Winning이 검증을 담당하도록 변경하여 컨트롤러의 역할을 명확히 함

jeongyou committed:
test(Winning): Validate 변경 사항 반영 및 테스트 코드 개선

- Validate 리팩터링 이후의 구조와 맞도록 테스트 코드를 개선
- Validate의 역할이 단순화됨에 따라, Winning이 직접 검증을 수행하는 구조로 변경
- Winning 클래스의 예외 처리 테스트 추가: 중복된 번호, 개수 초과, 범위 초과 등의 검증
- 테스트 독립성을 높이기 위해 beforeEach를 제거하고 개별 Winning 객체를 생성

jeongyou committed:
test(Validate): 유효성 검사 테스트 리팩터링 및 구조 개선

- validateWinningNumbers() 와validateBonusNumber() 제거 후, 개별 검증 함수 기반으로 테스트 재작성

jeongyou committed:
refactor(LottoUI): getBonusNumber()가 문자열을 반환하도록 수정

- getBonusNumber()에서 Number 변환을 제거하고, 문자열 값을 반환하도록 변경
- UI에서 숫자 변환을 수행하지 않고, Winning에서 처리할 수 있도록 함

jeongyou committed:
refactor(LottoGame): Winning의 보너스 번호 검증 책임 분리 및 입력값 검증 개선

- Winning 생성 시 보너스 번호를 즉시 받지 않고, setBonusNumber()를 별도로 호출하도록 변경
- Winning이 validateBonusNumber()를 직접 수행하도록 수정하여 검증 책임 분리
- getAndValidateWinningNumbers()에서 보너스 번호를 검증 후 숫자로 변환하도록 개선
- checkBonusNumberIsEmpty() 메서드 추가하여 보너스 번호의 빈 입력값 검증 추가

jeongyou committed:
refactor(LottoGame): LottoGameManager를 추가하여 상태 및 로직 분리

- LottoGame에서 price 및 lottos 상태를 직접 관리하지 않도록 구조 개선
- LottoGameManager를 추가하여 상태(price, lottos)와 로또 구매 로직을 전담하도록 변경
- LottoGame은 LottoGameManager를 사용하여 로직을 수행하고 UI 업데이트만 담당하도록 리팩터링
- Validate.validatePrice(price)를 LottoGameManager로 이동하여 검증 책임을 도메인 객체에서 수행하도록 개선

jeongyou committed:
fix: import 경로 수정하여 오류 해결

jeongyou committed:
chore: 파일 이름 변경

jeongyou committed:
refactor: LottoGame이 UI와 입력값 검증만 담당하도록 리팩토링

- LottoGame에서 직접 Winning 관련 로직을 수행하던 부분 제거
- LottoGameManager를 통해 당첨 검사 및 결과 계산 수행
- LottoGame은 UI 업데이트 및 데이터 흐름만 조율하는 역할로 변경
- 컨트롤러가 비대해지는 문제 해결

jeongyou commented:
안녕하세요, 동키콩! 점심 뭐 먹지 미션을 끝내느라 리뷰 요청이 너무 늦어져서 죄송합니다! 🙏🙏🙏

컨트롤러의 역할과 객체 간의 결합도에 대해 여러 차례 피드백을 주셔서 감사합니다! 솔직히 컨트롤러의 역할이 잘 이해가 안되었었는데 덕분에 컨트롤러가 가져야 할 책임을 더 명확하게 이해할 수 있었습니다.

리뷰에서 말씀해주신 것처럼, 컨트롤러가 도메인 로직을 직접 수행하지 않도록 LottoGameManager를 추가하여 상태와 로직을 분리했습니다. LottoGame은 UI와 데이터 흐름을 관리하는 역할만 수행하고, LottoGameManager가 당첨 검사 및 상태 관리를 담당하도록 개선했습니다.

**리팩토링 주요 변경 사항**
- 컨트롤러가 직접 도메인 로직을 수행하지 않도록 LottoGameManager 추가 → [621378a](https://github.com/woowacourse/javascript-lotto/pull/395/commits/621378aa0135998d8c151e9123214af201e216e4)
- Winning이 보너스 번호 검증을 직접 수행하도록 수정 → [d698856](https://github.com/woowacourse/javascript-lotto/pull/395/commits/d6988566c7a8ab2a1b8602c8a6c833a6438f5072)
- 컨트롤러가 UI와 데이터 흐름만 관리하도록 역할 정리 → [c65546e](https://github.com/woowacourse/javascript-lotto/pull/395/commits/c65546e2ef939218db97358735465f876ae22787)
- 커밋을 여러 번 나눠서 진행하다 보니, 변경 사항을 확인하기 어려울 수도 있을 것 같습니다. 혹시 커밋 로그로 확인하기 어렵다면, 전체 코드에서 변경된 부분을 봐주시면 감사하겠습니다! 🙏

❓추가로, 입력값에 대한 공백 검사는 Winning과 직접적인 연관이 없다고 판단하여 컨트롤러에서 처리하도록 남겨두었는데, 이 부분에 대해서는 어떻게 생각하시나요...?

마지막으로, 리팩토링 방향이 리뷰에서 의도하신 내용과 잘 맞는지 확인해주시면 감사하겠습니다!

JUDONGHYEOK reviewed (approved):
안녕하세요 기린! 기린과 재밌는 이야기를 코멘트로 나누다보니 핑퐁이 길어진 것 같아요. 제가 요새 공부하고 있는 내용과도 이어지는 내용들이라서요. 조금이나마 기린에게 도움이 될만한 내용이 되었길 바래봅니다.

개선해주신 내용들 확인했는데요. 이전보다 훨씬 객체간의 책임이 명확해진 느낌입니다. 조금더 참고하시면 좋을 사항들은 코멘트에 정리해놓았으니 확인 부탁드려요~

> 추가로, 입력값에 대한 공백 검사는 Winning과 직접적인 연관이 없다고 판단하여 컨트롤러에서 처리하도록 남겨두었는데, 이 부분에 대해서는 어떻게 생각하시나요...?

컨트롤러의 역할을 어떻게 정의하느냐에 따라 다를 것 같은데요. 저는 데이터 검증의 측면에서 컨트롤러는 두 계층의 검증을 "조율"하고 적절한 오류 처리를 하는 역할이지, 직접 검증 로직을 수행하는 것은 컨트롤러의 역할 즉 단일 책임 원칙에서 벗어난다고 생각해요. 공백검사를 어떻게 바라보냐 혹은 그 목적성에 따라 도메인에 위치할 수도, UI에 위치할 수도 있지만 컨트롤러는 개인적으로 그 역할을 벗어나는 것 같아요.

이전보다 구조도 깔끔해지고 미션의 요구사항도 모두 만족하므로 어프로브하도록 하겠습니다~~ 기린덕분에 재미있는 리뷰였네요. 로또게임 2단계 미션도 너무 수고많으셨어요~~
